cpp/src/arrow/compute/kernels/scalar_nested_test.cc

```diff
@@ -225,6 +225,167 @@ TEST(TestScalarNested, StructField) {
+      ["foo", 22]],
+      []
+    ]
+  )";
+  auto map_array = ArrayFromJSON(type, input);
+
+  CheckScalar("map_array_lookup", {map_array},
+              ArrayFromJSON(list(int32()), "[[99, 3], null, [101, 22], null]"), &foo_all);
+  CheckScalar("map_array_lookup", {map_array},
+              ArrayFromJSON(int32(), "[99, null, 101, null]"), &foo_first);
+  CheckScalar("map_array_lookup", {map_array},
+              ArrayFromJSON(int32(), "[3, null, 22, null]"), &foo_last);
+}
+
+TEST(TestScalarNested, MapArrayLookupNested) {
+  auto type = map(utf8(), map(int16(), int16()));
+  const char* input = R"(
+    [
+      [
+        [
+          "just",
+          [[0, 0], [1, 1]]
+        ],
+        [
+          "random",
+          [[2, 2], [3, 3]]
+        ],
+        [
+          "foo",
+          [[4, 4], [5, 5]]
+        ],
+        [
+          "values",
+          [[6, 6], [7, 7]]
+        ],
+        [
+          "foo",
+          [[8, 8], [9, 9]]
+        ],
+        [
+          "point",
+          [[10, 10], [11, 11]]
+        ],
+        [
+          "foo",
+          [[12, 12], [13, 13]]
+        ]
+      ],
+      null,
+      [
+        [
+          "yet",
+          [[0, 1], [1, 2]]
+        ],
+        [
+          "more",
+          [[2, 3], [3, 4]]
+        ],
+        [
+          "foo",
+          [[4, 5], [5, 6]]
+        ],
+        [
+          "random",
+          [[6, 7], [7, 8]]
+        ],
+        [
+          "foo",
+          [[8, 9], [9, 10]]
+        ],
+        [
+          "values",
+          [[10, 11], [11, 12]]
+        ],
+        [
+          "foo",
+          [[12, 13], [13, 14]]
+        ]
+      ],
+      []
+    ]
+  )";
+  auto map_array = ArrayFromJSON(type, input);
+
+  MapArrayLookupOptions foo_all(MakeScalar("foo"), MapArrayLookupOptions::ALL);
+  MapArrayLookupOptions foo_first(MakeScalar("foo"), MapArrayLookupOptions::FIRST);
+  MapArrayLookupOptions foo_last(MakeScalar("foo"), MapArrayLookupOptions::LAST);
+
+  auto foo_all_output = ArrayFromJSON(
+      list(map(int16(), int16())),
+      "[ [[[4, 4], [5, 5]], [[8, 8], [9, 9]], [[12, 12], [13, 13]]], null, [[[4, 5], [5, "
+      "6]], [[8, 9], [9, 10]], [[12, 13], [13, 14]]], null ]");
+
+  CheckScalar("map_array_lookup", {map_array}, foo_all_output, &foo_all);
+  CheckScalar("map_array_lookup", {map_array},
+              ArrayFromJSON(map(int16(), int16()),
+                            "[ [[4, 4], [5, 5]], null, [[4, 5], [5, 6]], null ]"),
+              &foo_first);
+  CheckScalar("map_array_lookup", {map_array},
+              ArrayFromJSON(map(int16(), int16()),
+                            "[ [[12, 12], [13, 13]], null, [[12, 13], [13, 14]], null ]"),
+              &foo_last);
+}
+
+template <typename Type>
+class TestMapArrayLookupIntegralKeys : public ::testing ::Test {};
+
+TYPED_TEST_SUITE(TestMapArrayLookupIntegralKeys, IntegralArrowTypes);
+
+TYPED_TEST(TestMapArrayLookupIntegralKeys, StringItems) {
```

Comment:
This basic structure has to be repeated for `StringArrowTypes` and `DecimalArrowTypes`, right? Also should templated tests be written for `BinaryArrowTypes` and `StringArrowTypes`?

Reply:
I would say these are the "groups" you need: `PhysicalIntegralArrowTypes`, `DecimalArrowTypes`, `BaseBinaryArrowTypes`, and then miscellaneous types (Boolean, FixedSizeString, MonthDayNanoInterval)

See here for all of the categories (note they overlap/not all are relevant): https://github.com/apache/arrow/blob/c39caedc2e9296fcb53d1c587708836ec6628ddb/cpp/src/arrow/testing/gtest_util.h#L153-L185